lib/internal/modules/esm/resolve.js

```diff
@@ -270,12 +270,17 @@ function finalizeResolution(resolved, base) {
       resolved.pathname, fileURLToPath(base), 'module');
   }
 
-  if (StringPrototypeEndsWith(resolved.pathname, '/')) return resolved;
   const path = fileURLToPath(resolved);
+  const stats = tryStatSync(path);
 
-  if (!tryStatSync(path).isFile()) {
-    throw new ERR_MODULE_NOT_FOUND(
+
+  if (!stats.isFile()) {
+    const err = new ERR_MODULE_NOT_FOUND(
       path || resolved.pathname, fileURLToPath(base), 'module');
+    if (stats.isDirectory()) {
+      err.directoryURL = resolved.toString();
```

Comment:
```suggestion
      err.directoryURL = String(resolved);
```
otherwise it's not actually guaranteed to be a string.